orttraining/orttraining/python/training/ortmodule/torch_cpp_extensions/_aten_op_factory.py

```diff
@@ -0,0 +1,124 @@
+# -------------------------------------------------------------------------
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# Licensed under the MIT License.
+# --------------------------------------------------------------------------
+
+"""Support registration of ATenOp's symbolic (PyTorch exporter overriding) and gradient definition.
+
+Each gradient definition of ATenOp is a list of node definitions: [node_def1, node_def2, ...].
+    Each node definition is a tuple: (op_type, inputs, outputs, attributes), while attributes is optional.
+        'op_type' is a string or a tuple of two strings with op_name and domain respectively.
+            If it's string type, then the domain is the default ONNX domain: ''.
+        'inputs' is a list of strings. Each input string can be GO(i), I(i), O(i),
+            which means ith gradient output, ith input, ith output of forward node respectively,
+            or any string for intermediate output of one of other nodes.
+        'outputs' is a list of strings. Each input string can be GI(i),
+            which means ith gradient input of forward node,
+            or any string for intermediate output for some of other nodes.
+        'attributes' (if present) is a dictionary. Each entry's key is the attribute name,
+            the value is also a dictionary: {'value': v, 'dtype': t, 'is_tensor': b}.
+                'v' can be string, number, or list of numbers, which is the value of the attribute.
+                't' is a string to describe the element type. It can be 'int', 'float', 'bool', etc,
+                    or 'IElemType(i)', 'OElemType(i)', which means the same type of ith input or
+                    output of forward node.
+                'b' is True or False, indicating if this attribute is tensor_proto ot not.
+                    'is_tensor' is optional, if not present, the default is False.
+"""
+
+import json
+from torch.onnx.symbolic_helper import _get_tensor_dim_size, _get_tensor_sizes
+
+
+def gradient_definition_to_json(gradient_def):
```

Comment:
I know json is an easy bridge between python structure-free strings to backend's parser, but I am wondering if we should introduce some structure to represent this gradient definition.

Reply:
Previously I though I needed the torch extension to pass by the definition, then string is the best as we didn't need to define structure on both ORT and torch extension. But on second thought, we don't need the torch extension for this. I've changed to use pybind for this.